### ISSUE TITLE
S1-04 [Backend] Fetch UCDP conflict data for all monitored countries

### ISSUE THREAD

**Issue by NathanLuckock** [Nathan7108/HackathonPractice#7]
## Problem
UCDP (Uppsala Conflict Data Program) provides the historical backbone for ML training — 75+ years of labeled conflict data back to 1946. Without it, the XGBoost risk scorer has no historical baseline and the training dataset is limited to ACLED's 1997-present coverage. UCDP is also the most academically cited conflict dataset on earth, which matters for judge credibility.

## Goal
Fetch UCDP Georeferenced Event Dataset (GED) and Armed Conflict Dataset for all monitored countries, save to `data/ucdp/`, expose `compute_ucdp_features()` returning 5 features, and `build_ucdp_training_labels()` for auto-labeling historical country-years.

### In scope
- `fetch_ucdp_ged(country_name)` → DataFrame via REST API (no key needed)
- `fetch_ucdp_armed_conflict()` → DataFrame via CSV download (no key needed)
- `compute_ucdp_features(ged_df, window_years=5)` → 5 features matching ML Guide Section 2.3
- `build_ucdp_training_labels(armed_conflict_df)` → labeled country-years for XGBoost training
- Save to `data/ucdp/{country}_ged.csv` and `data/ucdp/armed_conflict_1946_2023.csv`
- CLI runner: `python -m backend.ml.data.fetch_ucdp`

### Out of scope
- Integration with feature pipeline (S1-06)
- Model training (S2-02)

## Acceptance Criteria
- [ ] `fetch_ucdp_ged("Ukraine")` returns DataFrame with columns including `year`, `country`, `deaths_a`, `deaths_b`, `deaths_civilians`, `type_of_violence`, `latitude`, `longitude`
- [ ] Pagination handled — API returns paginated results, all pages fetched
- [ ] `fetch_ucdp_armed_conflict()` downloads full dataset (1946–2023) and saves to `data/ucdp/armed_conflict_1946_2023.csv`
- [ ] `compute_ucdp_features(ged_df)` returns dict with exactly 5 keys: `ucdp_total_deaths`, `ucdp_state_conflict_years`, `ucdp_civilian_deaths`, `ucdp_conflict_intensity`, `ucdp_recurrence_rate`
- [ ] `build_ucdp_training_labels(armed_conflict_df)` returns DataFrame with columns: `location`, `year`, `conflict_id`, `intensity_level`, `risk_label`
- [ ] Risk labels distributed across LOW/ELEVATED/HIGH (print value counts)
- [ ] All 8 country GED CSVs saved to `data/ucdp/`
- [ ] Running as module fetches everything and prints summary stats

### UCDP REST API — no key required:
```python
# GED (Georeferenced Event Dataset)
# Base URL: https://ucdpapi.pcr.uu.se/api/gedevents/23.1
# Params: Country=<name>, pagesize=1000, page=N
# Returns JSON: { "TotalCount": N, "Result": [...] }

# Armed Conflict Dataset — direct CSV download:
# https://ucdp.uu.se/downloads/ucdpprio/ucdp-prio-acd-241.csv
```

### Implementation from ML Guide Section 2.3:
```python
def fetch_ucdp_ged(country_name: str) -> pd.DataFrame:
    url = "https://ucdpapi.pcr.uu.se/api/gedevents/23.1"
    params = {"Country": country_name, "pagesize": 1000, "page": 1}
    all_results = []
    while True:
        response = requests.get(url, params=params, timeout=30)
        data = response.json()
        results = data.get('Result', [])
        if not results:
            break
        all_results.extend(results)
        if len(results) < params['pagesize']:
            break
        params['page'] += 1
    df = pd.DataFrame(all_results)
    if len(df) > 0:
        df.to_csv(f"data/ucdp/{country_name.lower()}_ged.csv", index=False)
    return df
```

### Feature computation (5 features):
```python
def compute_ucdp_features(ged_df: pd.DataFrame, window_years: int = 5) -> dict:
    if len(ged_df) == 0:
        return {k: 0 for k in [
            'ucdp_total_deaths', 'ucdp_state_conflict_years',
            'ucdp_civilian_deaths', 'ucdp_conflict_intensity',
            'ucdp_recurrence_rate'
        ]}
    ged_df['year'] = pd.to_numeric(ged_df.get('year', 0), errors='coerce')
    recent = ged_df[ged_df['year'] >= ged_df['year'].max() - window_years]
    state_based = recent[recent.get('type_of_violence', pd.Series()).astype(str) == '1']
    civilian = recent[recent.get('type_of_violence', pd.Series()).astype(str) == '3']
    return {
        'ucdp_total_deaths': float(
            recent.get('deaths_a', pd.Series(0)).sum() +
            recent.get('deaths_b', pd.Series(0)).sum()
        ),
        'ucdp_state_conflict_years': len(state_based['year'].unique()) if len(state_based) > 0 else 0,
        'ucdp_civilian_deaths': float(civilian.get('deaths_civilians', pd.Series(0)).sum()),
        'ucdp_conflict_intensity': float(recent.get('deaths_a', pd.Series(0)).mean()),
        'ucdp_recurrence_rate': float(len(recent['year'].unique()) / max(window_years, 1)),
    }
```

### Training label generation (critical for XGBoost):
```python
def build_ucdp_training_labels(armed_conflict_df: pd.DataFrame) -> pd.DataFrame:
    df = armed_conflict_df.copy()
    def assign_label(row):
        if pd.isna(row.get('conflict_id')):
            return 'LOW'
        intensity = row.get('intensity_level', 0)
        if intensity >= 2:
            return 'HIGH'
        if intensity == 1:
            return 'ELEVATED'
        return 'MODERATE'
    df['risk_label'] = df.apply(assign_label, axis=1)
    print(f"Labeled {len(df)} country-years")
    print(df['risk_label'].value_counts())
    return df[['location', 'year', 'conflict_id', 'intensity_level', 'risk_label']]
```

### Country name mapping (UCDP uses full names):
```python
COUNTRIES = {
    "UA": "Ukraine", "TW": "Taiwan", "IR": "Iran (Government of)",
    "VE": "Venezuela", "PK": "Pakistan", "ET": "Ethiopia",
    "RS": "Serbia", "BR": "Brazil",
}
```

### Edge cases:
- **Iran:** UCDP may use "Iran (Government of)" as the country name in the armed conflict dataset. Try "Iran" first, fall back to alternatives if empty.
- **Taiwan:** May have zero results in UCDP — handle gracefully with empty DataFrame.
- **Serbia:** May be listed under "Serbia (Yugoslavia)" historically — handle gracefully.
- **API version:** Using GED version 23.1. If this returns errors, try without version number.
- **Armed conflict CSV URL:** The exact URL may change with dataset version updates. If 241 fails, check https://ucdp.uu.se/downloads/ for the latest version number.
- **Hackathon tradeoff:** If the CSV download URL is broken, fall back to manually downloading from the UCDP website and placing in `data/ucdp/`. Note this in PROGRESS.md.

## Manual Test Steps
1. `python -m backend.ml.data.fetch_ucdp` — fetches all GED data + armed conflict dataset
2. Verify `data/ucdp/ukraine_ged.csv` exists and has rows
3. Verify `data/ucdp/armed_conflict_1946_2023.csv` exists and spans 1946–2023
4. `python -c "import pandas as pd; from backend.ml.data.fetch_ucdp import compute_ucdp_features; df=pd.read_csv('data/ucdp/ukraine_ged.csv'); print(compute_ucdp_features(df))"` — prints 5-key dict with non-zero deaths for Ukraine
5. `python -c "import pandas as pd; from backend.ml.data.fetch_ucdp import build_ucdp_training_labels; df=pd.read_csv('data/ucdp/armed_conflict_1946_2023.csv'); labels=build_ucdp_training_labels(df); print(labels['risk_label'].value_counts())"` — prints label distribution across LOW/ELEVATED/HIGH

---

## Cursor Agent Instructions

**Repo:** Nathan7108/HackathonPractice
**Working directory:** `C:\Users\natha\HackathonPractice`
**Branch:** `feature/s1-04-fetch-ucdp`

### STEP 0 — Read context
```bash
cat PROGRESS.md
```

### STEP 1 — Create branch
```bash
git checkout main && git pull origin main
git checkout -b feature/s1-04-fetch-ucdp
```

### STEP 2 — Implement
Replace placeholder `backend/ml/data/fetch_ucdp.py` with full implementation:
1. `fetch_ucdp_ged(country_name)` — REST API with pagination, save to CSV
2. `fetch_ucdp_armed_conflict()` — direct CSV download, save to `data/ucdp/armed_conflict_1946_2023.csv`
3. `compute_ucdp_features(ged_df, window_years=5)` — 5 features per ML Guide Section 2.3
4. `build_ucdp_training_labels(armed_conflict_df)` — label country-years as LOW/ELEVATED/HIGH
5. `COUNTRIES` dict mapping ISO2 → UCDP country names
6. `if __name__ == "__main__"` block that:
   - Fetches GED for all 8 countries with 1s sleep between
   - Downloads armed conflict dataset
   - Prints event counts per country
   - Prints sample `compute_ucdp_features()` for Ukraine
   - Prints training label distribution from `build_ucdp_training_labels()`

### STEP 3 — Validate
```bash
cd C:\Users\natha\HackathonPractice
python -m backend.ml.data.fetch_ucdp
```
Verify: GED CSVs in `data/ucdp/`, armed conflict CSV downloaded, label distribution printed.

### STEP 4 — Commit and push
```bash
git add backend/ml/data/fetch_ucdp.py
git commit -m "feat: S1-04 UCDP data fetcher — GED API + armed conflict dataset + training labels"
git push -u origin feature/s1-04-fetch-ucdp
```

### STEP 5 — Open PR
```bash
gh pr create --title "S1-04 [Backend] Fetch UCDP conflict data" --body "Closes #<THIS_ISSUE_NUMBER>" --base main
```

### STEP 6 — Update PROGRESS.md
Mark `S1-04: Fetch UCDP data` as complete.

### Definition of Done
- [ ] Branch created from main
- [ ] `fetch_ucdp.py` fully implemented
- [ ] All 8 country GED CSVs saved to `data/ucdp/`
- [ ] Armed conflict dataset downloaded and saved
- [ ] `compute_ucdp_features()` returns 5-key dict
- [ ] `build_ucdp_training_labels()` returns labeled DataFrame with value counts printed
- [ ] Pushed to origin
- [ ] PR created

### If You Get Stuck
- UCDP API docs → https://ucdp.uu.se/ and https://ucdpapi.pcr.uu.se/api/
- Feature spec → ML Guide v2 Section 2.3
- If GED API returns empty for a country: try alternate country name spellings
- If armed conflict CSV URL is dead: go to https://ucdp.uu.se/downloads/ and find the latest UCDP/PRIO Armed Conflict Dataset CSV link
- Training label logic → ML Guide v2 Section 2.3 `build_ucdp_training_labels`
